File: KoshiSolvers/SolverMethodHoina.cs
using System.Collections.Generic;
namespace KoshiSolvers
{
    public class HoinSolver : Solver
    {
        // Constructors
        public HoinSolver(string _Name, BehaviorOfSolver _Behaviour) : base(_Name, _Behaviour) { }

        // Methods
        public override List<Point> Solve(InitialValueProblem Task)
        {
            double StepSize = Task.H;
            double YIntermediate, FunctionIntermediateValue, FunctionValue;
            int i = 1;
            List<Point> Solution = new List<Point>();

            Solution.Add(new Point(Task.T0, Task.Y0));
            FunctionValue = Task.CountFunctionValue(Solution[0].X, Solution[0].Y);

            while (Solution[i - 1].X < Task.T)
            {
                double temp = Solution[i - 1].X + StepSize;

                YIntermediate = Solution[i - 1].Y + StepSize * FunctionValue;
                FunctionIntermediateValue = Task.CountFunctionValue(temp - StepSize, YIntermediate);
                Solution.Add(new Point(temp, Solution[i - 1].Y + StepSize / 2 * (FunctionValue + FunctionIntermediateValue)));

                FunctionValue = Task.CountFunctionValue(Solution[i].X, Solution[i].Y);

                if ((Solution[i].X + StepSize) > Task.T)
                {
                    // Out of border situations
                    switch (Behavior) 
                    {
                        case BehaviorOfSolver.FinishAtTheLeftBorder:
                            StepSize = Task.T - (StepSize * i + Solution[0].X);
                            temp = Solution[i].X + StepSize;

                            YIntermediate = Solution[i].Y + StepSize * FunctionValue;
                            FunctionIntermediateValue = Task.CountFunctionValue(Solution[i + 1].X, YIntermediate);
                            Solution.Add(new Point(temp, Solution[i].Y + StepSize / 2 * (FunctionValue + FunctionIntermediateValue)));
                            break;

                        case BehaviorOfSolver.FinishAfterLeftBorder:
                            temp = Solution[i].X + StepSize;
                            
                            YIntermediate = Solution[i].Y + StepSize * FunctionValue;
                            FunctionIntermediateValue = Task.CountFunctionValue(Solution[i + 1].X, YIntermediate);
                            Solution.Add(new Point(temp, Solution[i].Y + StepSize / 2 * (FunctionValue + FunctionIntermediateValue)));
                            break;
                        case BehaviorOfSolver.FinishBeforeLeftBorder:
                            break;
                    }
                    break;
                }
                i++;
            }
            return Solution;
        }
    }
}

File: C_Plus_Plus_Projects--2-4.1/SolverMethodHoina.cpp
#include "SolverMethodHoina.h"
#include <iostream>

void SolverMethodHoina::SolverKoshiTask(const TaskKoshi& Task)
{
	double StepSize = Task.Geth();
	double YIntermedied, FunctionIntermediedValue, FunctionValue;
	unsigned int i = 1; 

	Point.X[0] = Task.Gett0(); // Задание первоначальных значений
	Point.Y[0] = Task.Gety0();
	FunctionValue = Task.CountFunctionValue(Point.X[0], Point.Y[0]);


	while (Point.X[i] <= Task.GetT())
	{
		Point.X[i] = Point.X[i - 1] + StepSize;
		YIntermedied = Point.Y[i - 1] + StepSize * FunctionValue;
		FunctionIntermediedValue = Task.CountFunctionValue(Point.X[i], YIntermedied);
		Point.Y[i] = Point.Y[i - 1] + StepSize / 2 * (FunctionValue + FunctionIntermediedValue);
		FunctionValue = Task.CountFunctionValue(Point.X[i], Point.Y[i]);


		// Особый случай не поподания на конец интервала
		if ((Point.X[i] + StepSize) > Task.GetT())
		{
			switch (Behavior)//// // случаи непопадания на границу отрезка
			{
			case BehaviorOfSolver::NoneBehavior:
				break;

			case BehaviorOfSolver::FinishAtTheLeftBorder: 

				StepSize = Task.GetT() - (StepSize * i + Point.X[0]);
				
				Point.X[i + 1] = Point.X[i] + StepSize;
				YIntermedied = Point.Y[i] + StepSize * FunctionValue;
				FunctionIntermediedValue = Task.CountFunctionValue(Point.X[i + 1], YIntermedied);
				Point.Y[i + 1] = Point.Y[i] + StepSize / 2 * (FunctionValue + FunctionIntermediedValue);
				
				break;
			case BehaviorOfSolver::FinishAfterLeftBorder: 

		
				Point.X[i + 1] = Point.X[i] + StepSize;
				YIntermedied = Point.Y[i] + StepSize * FunctionValue;
				FunctionIntermediedValue = Task.CountFunctionValue(Point.X[i + 1], YIntermedied);
				Point.Y[i + 1] = Point.Y[i] + StepSize / 2 * (FunctionValue + FunctionIntermediedValue);
	
				break;
			case BehaviorOfSolver::FinishBeforeLeftBorder: 
				break;
			}
			break; 
		}
		i++;
	}
}
